File: app/src/main/java/com/krystianwsul/checkme/gui/friends/FriendListFragment.java
package com.krystianwsul.checkme.gui.friends;


import android.content.Context;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.Loader;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.view.ActionMode;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.krystianwsul.checkme.R;
import com.krystianwsul.checkme.domainmodel.DomainFactory;
import com.krystianwsul.checkme.gui.AbstractFragment;
import com.krystianwsul.checkme.gui.FabUser;
import com.krystianwsul.checkme.gui.SelectionCallback;
import com.krystianwsul.checkme.loaders.FriendListLoader;

import junit.framework.Assert;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;

public class FriendListFragment extends AbstractFragment implements LoaderManager.LoaderCallbacks<FriendListLoader.Data>, FabUser {
    private static final String SELECTED_IDS_KEY = "selectedIds";

    private ProgressBar mFriendListProgress;
    private RecyclerView mFriendListRecycler;
    private TextView mEmptyText;

    private FriendListAdapter mFriendListAdapter;

    @Nullable
    private FriendListLoader.Data mData;

    @Nullable
    private List<String> mSelectedIds;

    private final SelectionCallback mSelectionCallback = new SelectionCallback() {
        @Override
        protected void unselect() {
            mFriendListAdapter.unselect();
        }

        @Override
        protected void onMenuClick(MenuItem menuItem) {
            List<String> selectedUserDataEmails = mFriendListAdapter.getSelected();
            Assert.assertTrue(!selectedUserDataEmails.isEmpty());

            switch (menuItem.getItemId()) {
                case R.id.action_custom_times_delete:
                    mFriendListAdapter.removeSelected();

                    updateSelectAll();

                    break;
                default:
                    throw new UnsupportedOperationException();
            }
        }

        @Override
        protected void onFirstAdded() {
            ((AppCompatActivity) getActivity()).startSupportActionMode(this);

            mActionMode.getMenuInflater().inflate(R.menu.menu_custom_times, mActionMode.getMenu());

            updateFabVisibility();

            ((Listener) getActivity()).onCreateUserActionMode(mActionMode);
        }

        @Override
        protected void onSecondAdded() {

        }

        @Override
        protected void onOtherAdded() {

        }

        @Override
        protected void onLastRemoved() {
            updateFabVisibility();

            ((Listener) getActivity()).onDestroyUserActionMode();
        }

        @Override
        protected void onSecondToLastRemoved() {

        }

        @Override
        protected void onOtherRemoved() {

        }
    };

    @Nullable
    private FloatingActionButton mFriendListFab;

    @NonNull
    public static FriendListFragment newInstance() {
        return new FriendListFragment();
    }

    public FriendListFragment() {

    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        Assert.assertTrue(context instanceof Listener);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_friend_list, container, false);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        RelativeLayout friendListLayout = (RelativeLayout) getView();
        Assert.assertTrue(friendListLayout != null);

        mFriendListProgress = (ProgressBar) friendListLayout.findViewById(R.id.friend_list_progress);
        Assert.assertTrue(mFriendListProgress != null);

        mFriendListRecycler = (RecyclerView) friendListLayout.findViewById(R.id.friend_list_recycler);
        Assert.assertTrue(mFriendListRecycler != null);

        mFriendListRecycler.setLayoutManager(new LinearLayoutManager(getActivity()));

        mEmptyText = (TextView) friendListLayout.findViewById(R.id.empty_text);
        Assert.assertTrue(mEmptyText != null);

        if (savedInstanceState != null && savedInstanceState.containsKey(SELECTED_IDS_KEY)) {
            mSelectedIds = savedInstanceState.getStringArrayList(SELECTED_IDS_KEY);
            Assert.assertTrue(mSelectedIds != null);
        }

        getLoaderManager().initLoader(0, null, this);
    }

    @Override
    public Loader<FriendListLoader.Data> onCreateLoader(int id, Bundle args) {
        return new FriendListLoader(getActivity());
    }

    @Override
    public void onLoadFinished(Loader<FriendListLoader.Data> loader, FriendListLoader.Data data) {
        Assert.assertTrue(data != null);

        mData = data;

        if (mFriendListAdapter != null)
            mSelectedIds = mFriendListAdapter.getSelected();
        else if (mSelectedIds == null)
            mSelectedIds = new ArrayList<>();

        mFriendListAdapter = new FriendListAdapter(data.mUserListDatas, mSelectedIds);
        mFriendListRecycler.setAdapter(mFriendListAdapter);

        mSelectionCallback.setSelected(mFriendListAdapter.getSelected().size());

        mFriendListProgress.setVisibility(View.GONE);

        updateFabVisibility();

        if (data.mUserListDatas.isEmpty()) {
            mFriendListRecycler.setVisibility(View.GONE);
            mEmptyText.setVisibility(View.VISIBLE);
            mEmptyText.setText(R.string.friends_empty);
        } else {
            mFriendListRecycler.setVisibility(View.VISIBLE);
            mEmptyText.setVisibility(View.GONE);
        }

        updateSelectAll();
    }

    @Override
    public void onLoaderReset(Loader<FriendListLoader.Data> loader) {
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        if (mFriendListAdapter != null)
            outState.putStringArrayList(SELECTED_IDS_KEY, new ArrayList<>(mFriendListAdapter.getSelected()));
    }

    private void updateSelectAll() {
        Assert.assertTrue(mFriendListAdapter != null);

        ((Listener) getActivity()).setUserSelectAllVisibility(mFriendListAdapter.getItemCount() != 0);
    }

    public void selectAll() {
        mFriendListAdapter.selectAll();
    }

    @Override
    public void setFab(@NonNull FloatingActionButton floatingActionButton) {
        Assert.assertTrue(mFriendListFab == null);

        mFriendListFab = floatingActionButton;

        mFriendListFab.setOnClickListener(v -> startActivity(FindFriendActivity.newIntent(getActivity())));

        updateFabVisibility();
    }

    private void updateFabVisibility() {
        if (mFriendListFab == null)
            return;

        if (mData != null && !mSelectionCallback.hasActionMode()) {
            mFriendListFab.show();
        } else {
            mFriendListFab.hide();
        }
    }

    @Override
    public void clearFab() {
        if (mFriendListFab == null)
            return;

        mFriendListFab.setOnClickListener(null);

        mFriendListFab.hide();

        mFriendListFab = null;
    }

    public class FriendListAdapter extends RecyclerView.Adapter<FriendListAdapter.FriendHolder> {
        @NonNull
        private final List<UserDataWrapper> mUserDataWrappers;

        FriendListAdapter(@NonNull Collection<FriendListLoader.UserListData> userListDatas, @NonNull List<String> selectedIds) {
            Assert.assertTrue(mData != null);

            Map<String, FriendListLoader.UserListData> userListMap = Stream.of(userListDatas)
                    .collect(Collectors.toMap(userListData -> userListData.mId, userListData -> userListData));

            mUserDataWrappers = Stream.of(userListMap.values())
                    .sorted((lhs, rhs) -> lhs.mId.compareTo(rhs.mId))
                    .map(userListData -> new UserDataWrapper(userListData, selectedIds))
                    .collect(Collectors.toList());
        }

        @Override
        public int getItemCount() {
            return mUserDataWrappers.size();
        }

        void unselect() {
            Stream.of(mUserDataWrappers)
                    .filter(userDataWrapper -> userDataWrapper.mSelected)
                    .forEach(userDataWrapper -> {
                        userDataWrapper.mSelected = false;
                        notifyItemChanged(mUserDataWrappers.indexOf(userDataWrapper));
                    });
        }

        @Override
        public FriendHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            LayoutInflater layoutInflater = LayoutInflater.from(getActivity());
            View friendRow = layoutInflater.inflate(R.layout.row_friend, parent, false);

            TextView friendName = (TextView) friendRow.findViewById(R.id.friend_name);
            Assert.assertTrue(friendName != null);

            TextView friendEmail = (TextView) friendRow.findViewById(R.id.friend_email);
            Assert.assertTrue(friendEmail != null);

            return new FriendHolder(friendRow, friendName, friendEmail);
        }

        @Override
        public void onBindViewHolder(final FriendHolder friendHolder, int position) {
            UserDataWrapper userDataWrapper = mUserDataWrappers.get(position);
            Assert.assertTrue(userDataWrapper != null);

            friendHolder.mFriendName.setText(userDataWrapper.mUserListData.mName);
            friendHolder.mFriendEmail.setText(userDataWrapper.mUserListData.mEmail);

            if (userDataWrapper.mSelected)
                friendHolder.mFriendRow.setBackgroundColor(ContextCompat.getColor(getActivity(), R.color.selected));
            else
                friendHolder.mFriendRow.setBackgroundColor(Color.TRANSPARENT);

            friendHolder.mFriendRow.setOnLongClickListener(v -> {
                friendHolder.onLongClick();
                return true;
            });

            friendHolder.mFriendRow.setOnClickListener(v -> {
                if (mSelectionCallback.hasActionMode())
                    friendHolder.onLongClick();
                else
                    friendHolder.onRowClick();
            });
        }

        @NonNull
        List<String> getSelected() {
            return new ArrayList<>(Stream.of(mUserDataWrappers)
                    .filter(userDataWrapper -> userDataWrapper.mSelected)
                    .map(userDataWrapper -> userDataWrapper.mUserListData.mId)
                    .collect(Collectors.toList()));
        }

        void removeSelected() {
            List<UserDataWrapper> selectedUserDataWrappers = Stream.of(mUserDataWrappers)
                    .filter(userDataWrapper -> userDataWrapper.mSelected)
                    .collect(Collectors.toList());

            for (UserDataWrapper userDataWrapper : selectedUserDataWrappers) {
                int position = mUserDataWrappers.indexOf(userDataWrapper);
                mUserDataWrappers.remove(position);
                notifyItemRemoved(position);
            }

            DomainFactory.getDomainFactory(getActivity())
                    .removeFriends(Stream.of(selectedUserDataWrappers)
                            .map(userDataWrapper -> userDataWrapper.mUserListData.mId)
                            .collect(Collectors.toSet()));
        }

        public void selectAll() {
            Assert.assertTrue(!mSelectionCallback.hasActionMode());

            Stream.of(mUserDataWrappers)
                    .filter(customTimeWrapper -> !customTimeWrapper.mSelected)
                    .forEach(UserDataWrapper::toggleSelect);
        }

        class FriendHolder extends RecyclerView.ViewHolder {
            final View mFriendRow;
            final TextView mFriendName;
            final TextView mFriendEmail;

            FriendHolder(@NonNull View friendRow, @NonNull TextView friendName, @NonNull TextView friendEmail) {
                super(friendRow);

                mFriendRow = friendRow;
                mFriendName = friendName;
                mFriendEmail = friendEmail;
            }

            void onRowClick() {
                UserDataWrapper userDataWrapper = mUserDataWrappers.get(getAdapterPosition());
                Assert.assertTrue(userDataWrapper != null);
            }

            void onLongClick() {
                int position = getAdapterPosition();

                UserDataWrapper userDataWrapper = mUserDataWrappers.get(position);
                Assert.assertTrue(userDataWrapper != null);

                userDataWrapper.toggleSelect();
            }
        }
    }

    private class UserDataWrapper {
        @NonNull
        final FriendListLoader.UserListData mUserListData;

        boolean mSelected = false;

        UserDataWrapper(@NonNull FriendListLoader.UserListData userListData, @NonNull List<String> selectedIds) {
            mUserListData = userListData;

            mSelected = selectedIds.contains(mUserListData.mId);
        }

        void toggleSelect() {
            mSelected = !mSelected;

            if (mSelected) {
                mSelectionCallback.incrementSelected();
            } else {
                mSelectionCallback.decrementSelected();
            }

            int position = mFriendListAdapter.mUserDataWrappers.indexOf(this);
            Assert.assertTrue(position >= 0);

            mFriendListAdapter.notifyItemChanged(position);
        }
    }

    public interface Listener {
        void onCreateUserActionMode(@NonNull ActionMode actionMode);

        void onDestroyUserActionMode();

        void setUserSelectAllVisibility(boolean selectAllVisible);
    }

}

File: app/src/main/java/com/krystianwsul/checkme/gui/friends/FriendListFragment.kt
package com.krystianwsul.checkme.gui.friends


import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.CustomItemAnimator
import com.jakewharton.rxrelay3.BehaviorRelay
import com.krystianwsul.checkme.R
import com.krystianwsul.checkme.databinding.FragmentFriendListBinding
import com.krystianwsul.checkme.domainmodel.extensions.addFriends
import com.krystianwsul.checkme.domainmodel.extensions.removeFriends
import com.krystianwsul.checkme.domainmodel.update.AndroidDomainUpdater
import com.krystianwsul.checkme.gui.base.AbstractFragment
import com.krystianwsul.checkme.gui.base.ActionModeListener
import com.krystianwsul.checkme.gui.base.SnackbarListener
import com.krystianwsul.checkme.gui.friends.findfriend.FindFriendActivity
import com.krystianwsul.checkme.gui.main.FabUser
import com.krystianwsul.checkme.gui.main.MainActivity
import com.krystianwsul.checkme.gui.tree.AbstractHolder
import com.krystianwsul.checkme.gui.tree.AbstractModelNode
import com.krystianwsul.checkme.gui.tree.BaseAdapter
import com.krystianwsul.checkme.gui.tree.HolderType
import com.krystianwsul.checkme.gui.tree.delegates.avatar.AvatarDelegate
import com.krystianwsul.checkme.gui.tree.delegates.avatar.AvatarModelNode
import com.krystianwsul.checkme.gui.tree.delegates.multiline.MultiLineDelegate
import com.krystianwsul.checkme.gui.tree.delegates.multiline.MultiLineModelNode
import com.krystianwsul.checkme.gui.tree.delegates.multiline.MultiLineRow
import com.krystianwsul.checkme.gui.utils.BottomFabMenuDelegate
import com.krystianwsul.checkme.gui.utils.ResettableProperty
import com.krystianwsul.checkme.gui.utils.SelectionCallback
import com.krystianwsul.checkme.gui.widgets.MyBottomBar
import com.krystianwsul.checkme.utils.animateVisibility
import com.krystianwsul.checkme.viewmodels.FriendListViewModel
import com.krystianwsul.checkme.viewmodels.getViewModel
import com.krystianwsul.common.utils.UserKey
import com.krystianwsul.treeadapter.*
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.kotlin.addTo
import io.reactivex.rxjava3.kotlin.plusAssign
import java.util.*

class FriendListFragment : AbstractFragment(), FabUser {

    companion object {

        private const val SELECTED_IDS_KEY = "selectedIds"

        fun newInstance() = FriendListFragment()
    }

    lateinit var treeViewAdapter: TreeViewAdapter<AbstractHolder>
        private set

    private var data: FriendListViewModel.Data? = null

    private var selectedIds = listOf<UserKey>()

    private val listener get() = activity as FriendListListener

    private val selectionCallback = object : SelectionCallback() {

        override val activity get() = requireActivity()

        override fun getTreeViewAdapter() = treeViewAdapter

        override fun unselect(placeholder: TreeViewAdapter.Placeholder) = treeViewAdapter.unselect(placeholder)

        override val bottomBarData by lazy { Triple(listener.getBottomBar(), R.menu.menu_friends, listener::initBottomBar) }

        override fun onMenuClick(itemId: Int, placeholder: TreeViewAdapter.Placeholder): Boolean {
            val selectedUserDataEmails = treeViewAdapter.selectedNodes
            check(selectedUserDataEmails.isNotEmpty())

            when (itemId) {
                R.id.action_friends_delete -> {
                    (treeViewAdapter.treeModelAdapter as FriendListAdapter).removeSelected(placeholder)

                    updateSelectAll()
                }
                else -> throw UnsupportedOperationException()
            }

            return true
        }

        override fun onFirstAdded(placeholder: TreeViewAdapter.Placeholder, initial: Boolean) {
            (activity as AppCompatActivity).startSupportActionMode(this)

            updateFabVisibility()

            (activity as MainActivity).onCreateActionMode(actionMode!!)

            super.onFirstAdded(placeholder, initial)
        }

        override fun onLastRemoved(placeholder: TreeViewAdapter.Placeholder) {
            updateFabVisibility()

            (activity as MainActivity).onDestroyActionMode()
        }
    }

    private var friendListFabDelegate: BottomFabMenuDelegate.FabDelegate? = null

    private val isVisible = BehaviorRelay.createDefault(false)
    private lateinit var friendListViewModel: FriendListViewModel

    private val mainActivity get() = activity as MainActivity

    private val bindingProperty = ResettableProperty<FragmentFriendListBinding>()
    private var binding by bindingProperty

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) =
            FragmentFriendListBinding.inflate(inflater, container, false).also { binding = it }.root

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        if (savedInstanceState?.containsKey(SELECTED_IDS_KEY) == true)
            selectedIds = savedInstanceState.getParcelableArrayList(SELECTED_IDS_KEY)!!

        friendListViewModel = getViewModel<FriendListViewModel>().apply {
            viewCreatedDisposable += isVisible.subscribe { if (it) start() else stop() }

            viewCreatedDisposable += data.subscribe { onLoadFinished(it) }
        }
    }

    private fun getSelectedIds() = treeViewAdapter.selectedNodes.map { (it.modelNode as FriendNode).userListData.id }

    private fun onLoadFinished(data: FriendListViewModel.Data) {
        this.data = data

        if (this::treeViewAdapter.isInitialized) {
            selectedIds = getSelectedIds()

            treeViewAdapter.updateDisplayedNodes {
                (treeViewAdapter.treeModelAdapter as FriendListAdapter).initialize()

                selectionCallback.setSelected(treeViewAdapter.selectedNodes.size, it, false)
            }
        } else {
            val friendListAdapter = FriendListAdapter()
            friendListAdapter.initialize()
            treeViewAdapter = friendListAdapter.treeViewAdapter

            binding.friendListRecycler.apply {
                adapter = treeViewAdapter
                itemAnimator = CustomItemAnimator()
            }

            treeViewAdapter.updateDisplayedNodes {
                selectionCallback.setSelected(treeViewAdapter.selectedNodes.size, it, true)
            }
        }

        updateFabVisibility()

        val hide = mutableListOf<View>(binding.friendListProgress)
        val show: View

        if (data.userListDatas.isEmpty()) {
            hide += binding.friendListRecycler
            show = binding.friendListEmptyTextInclude.emptyTextLayout

            binding.friendListEmptyTextInclude
                    .emptyText
                    .setText(R.string.friends_empty)
        } else {
            show = binding.friendListRecycler
            hide += binding.friendListEmptyTextInclude.emptyTextLayout
        }

        animateVisibility(listOf(show), hide)

        updateSelectAll()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        if (this::treeViewAdapter.isInitialized)
            treeViewAdapter.let { outState.putParcelableArrayList(SELECTED_IDS_KEY, ArrayList(getSelectedIds())) }
    }

    private fun updateSelectAll() =
        (activity as MainActivity).setUserSelectAllVisibility(treeViewAdapter.displayedNodes.isNotEmpty())

    override fun setFab(fabDelegate: BottomFabMenuDelegate.FabDelegate) {
        friendListFabDelegate = fabDelegate

        friendListFabDelegate!!.setOnClickListener { startActivity(FindFriendActivity.newIntent(requireActivity())) }

        updateFabVisibility()

        isVisible.accept(true)
    }

    private fun updateFabVisibility() {
        friendListFabDelegate?.run {
            if (data != null && !selectionCallback.hasActionMode) {
                show()
            } else {
                hide()
            }
        }
    }

    override fun clearFab() {
        isVisible.accept(false)

        friendListFabDelegate = null
    }

    override fun onDestroyView() {
        bindingProperty.reset()

        super.onDestroyView()
    }

    private inner class FriendListAdapter : BaseAdapter(), ActionModeCallback by selectionCallback {

        lateinit var userDataWrappers: MutableList<FriendNode>

        val treeViewAdapter = TreeViewAdapter(
                this,
                TreeViewAdapter.PaddingData(R.layout.row_group_list_fab_padding, R.id.paddingProgress),
        )

        override lateinit var treeNodeCollection: TreeNodeCollection<AbstractHolder>
            private set

        fun initialize() {
            treeNodeCollection = TreeNodeCollection(treeViewAdapter)
            treeViewAdapter.setTreeNodeCollection(treeNodeCollection)

            userDataWrappers = data!!.userListDatas
                    .asSequence()
                    .sortedBy { it.id }
                    .map { FriendNode(it) }
                    .toMutableList()

            treeNodeCollection.nodes = userDataWrappers.map { it.initialize(treeNodeCollection) }
        }

        fun removeSelected(@Suppress("UNUSED_PARAMETER") placeHolder: TreeViewAdapter.Placeholder) {
            val selectedUserDataWrappers = userDataWrappers.filter { it.treeNode.isSelected }
            val userListDatas = selectedUserDataWrappers.map { it.userListData }
            val userPairs = userListDatas.associate { it.id to it.userWrapper }
            val friendIds = userPairs.map { it.key }.toSet()

            AndroidDomainUpdater.removeFriends(friendListViewModel.dataId.toFirst(), friendIds)
                    .observeOn(AndroidSchedulers.mainThread())
                    .andThen(mainActivity.showSnackbarRemovedMaybe(userListDatas.size))
                    .flatMapCompletable { AndroidDomainUpdater.addFriends(friendListViewModel.dataId.toFirst(), userPairs) }
                    .subscribe()
                    .addTo(createDisposable)
        }
    }

    private inner class FriendNode(val userListData: FriendListViewModel.UserListData) :
            AbstractModelNode(),
            AvatarModelNode,
            MultiLineModelNode {

        override val rowsDelegate = object : MultiLineModelNode.RowsDelegate {

            private val name = MultiLineRow.Visible(userListData.name)
            private val details = MultiLineRow.Visible(userListData.email, R.color.textSecondary)

            override fun getRows(isExpanded: Boolean, allChildren: List<TreeNode<*>>) = listOf(name, details)
        }

        public override lateinit var treeNode: TreeNode<AbstractHolder>
            private set

        override val holderType = HolderType.AVATAR

        override val isSelectable = true

        override val id = userListData.id

        override val parentNode: ModelNode<AbstractHolder>? = null

        override val avatarUrl = userListData.photoUrl

        override val delegates by lazy {
            listOf(
                AvatarDelegate(this),
                MultiLineDelegate(this), // this one always has to be last, because it depends on layout changes from prev
            )
        }

        override val widthKey
            get() = MultiLineDelegate.WidthKey(
                    0,
                    true,
                    false,
                    false
            )

        override fun compareTo(other: ModelNode<AbstractHolder>) = userListData.id.compareTo((other as FriendNode).userListData.id)

        fun initialize(treeNodeCollection: TreeNodeCollection<AbstractHolder>): TreeNode<AbstractHolder> {
            treeNode = TreeNode(
                    this,
                    treeNodeCollection,
                    selectedIds.contains(userListData.id),
            )

            treeNode.setChildTreeNodes(listOf())
            return treeNode
        }
    }

    interface FriendListListener : SnackbarListener, ActionModeListener {

        fun setUserSelectAllVisibility(selectAllVisible: Boolean)

        fun getBottomBar(): MyBottomBar

        fun initBottomBar()
    }
}
